symbolic-uplc/src/ZkFold/Symbolic/UPLC/Class.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE AllowAmbiguousTypes    #-}
+{-# LANGUAGE ConstraintKinds        #-}
+{-# LANGUAGE DataKinds              #-}
+{-# LANGUAGE FlexibleContexts       #-}
+{-# LANGUAGE FlexibleInstances      #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GADTs                  #-}
+{-# LANGUAGE KindSignatures         #-}
+{-# LANGUAGE RankNTypes             #-}
+{-# LANGUAGE ScopedTypeVariables    #-}
+{-# LANGUAGE TypeApplications       #-}
+{-# LANGUAGE TypeOperators          #-}
+{-# LANGUAGE UndecidableInstances   #-}
+
+
+module ZkFold.Symbolic.UPLC.Class (IsData (..), Sym, ExValue (..), ExList (..), BSLength, StrLength, IntLength) where
+
+import           Data.Functor.Rep                   (Representable)
+import           Data.Maybe                         (Maybe (..))
+import           Data.Proxy                         (Proxy (..))
+import           Data.Traversable                   (Traversable)
+import           Data.Typeable                      (Typeable)
+import           Prelude                            (type (~))
+
+import           ZkFold.Symbolic.Data.Bool          (Bool)
+import           ZkFold.Symbolic.Data.Class         (SymbolicData (..), SymbolicOutput)
+import           ZkFold.Symbolic.Data.Combinators
+import           ZkFold.Symbolic.Data.Conditional   (Conditional)
+import           ZkFold.Symbolic.Data.Int
+import qualified ZkFold.Symbolic.Data.List          as L
+import           ZkFold.Symbolic.Data.VarByteString
+import           ZkFold.Symbolic.Fold               (SymbolicFold)
+import qualified ZkFold.Symbolic.UPLC.Data          as Symbolic
+import           ZkFold.UPLC.BuiltinType
+
+
+-- | Class of Symbolic datatypes used inside Converter.
+-- Each instance enforces a one-to-one correspondence between some 'BuiltinType'
+-- and its interpretation as a Symbolic datatype in arbitrary context 'c'.
+class
+    ( Typeable v
+    , SymbolicOutput v, SymbolicFold c
+    , Context v ~ c, Support v ~ Proxy c
+    -- TODO: Remove after Conditional becomes part of SymbolicData
+    , Conditional (Bool c) v
+    , Representable (Layout v)
```

Comment:
Btw, `Representable` and `Traversable` constraints can be removed since these are already part of `SymbolicData`